File: internal/sync/tools.go
package sync

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/iskry/cask/internal/config"
	"github.com/iskry/cask/internal/executor"
)

// ToolsSyncManager synchronizes mise-managed tools between host and config.
type ToolsSyncManager struct {
	Config *config.ToolsConfig
}

func (m *ToolsSyncManager) ResourceType() string { return "tools" }

func (m *ToolsSyncManager) GetHostResources(ctx context.Context, exec executor.Executor) ([]Resource, error) {
	r, err := exec.Execute(ctx, []string{"mise", "list", "--json"})
	if err != nil || !r.Success || strings.TrimSpace(r.Stdout) == "" {
		return nil, nil
	}

	var data map[string]any
	if err := json.Unmarshal([]byte(r.Stdout), &data); err != nil {
		return nil, nil
	}

	var resources []Resource
	for toolName, versions := range data {
		switch v := versions.(type) {
		case []any:
			for _, ver := range v {
				if verMap, ok := ver.(map[string]any); ok {
					version, _ := verMap["version"].(string)
					resources = append(resources, Resource{
						"name":    toolName,
						"version": version,
					})
				}
			}
		case map[string]any:
			version, _ := v["version"].(string)
			resources = append(resources, Resource{
				"name":    toolName,
				"version": version,
			})
		}
	}
	return resources, nil
}

func (m *ToolsSyncManager) GetConfigResources() []Resource {
	if m.Config == nil {
		return nil
	}
	var resources []Resource
	for _, t := range m.Config.Tools {
		resources = append(resources, Resource{
			"name":           t.Name,
			"version":        t.Version,
			"global_install": t.GlobalInstall,
		})
	}
	return resources
}

func (m *ToolsSyncManager) ResourceID(r Resource) string {
	name, _ := r["name"].(string)
	return name
}

func (m *ToolsSyncManager) Apply(ctx context.Context, exec executor.Executor, r Resource) bool {
	name, _ := r["name"].(string)
	version, _ := r["version"].(string)
	if version == "" {
		version = "latest"
	}
	spec := fmt.Sprintf("%s@%s", name, version)

	result, err := exec.Execute(ctx, []string{"mise", "install", spec})
	if err != nil || !result.Success {
		return false
	}

	if globalInstall, _ := r["global_install"].(bool); globalInstall {
		result, err = exec.Execute(ctx, []string{"mise", "use", "--global", spec})
		return err == nil && result.Success
	}
	return true
}

func (m *ToolsSyncManager) Remove(ctx context.Context, exec executor.Executor, id string) bool {
	r, err := exec.Execute(ctx, []string{"mise", "uninstall", "--all", id})
	return err == nil && r.Success
}

func (m *ToolsSyncManager) NeedsUpdate(host, cfg Resource) bool {
	cfgVersion, _ := cfg["version"].(string)
	if cfgVersion == "latest" || cfgVersion == "stable" {
		return false
	}
	hostVersion, _ := host["version"].(string)
	return !strings.HasPrefix(hostVersion, cfgVersion)
}

File: src/cask/sync/tools.py
"""Mise tool sync implementation."""
from __future__ import annotations

from dataclasses import dataclass

from cask.executor.protocol import Executor
from cask.managers.mise import MiseManager
from cask.result import Result


@dataclass
class ToolResource:
    name: str
    version: str


class ToolSync:
    def __init__(self) -> None:
        self._mgr = MiseManager()

    async def get_host_resources(self, exec: Executor) -> list[ToolResource]:
        tools = await self._mgr.list_tools(exec)
        return [ToolResource(name=n, version=v) for n, v in tools.items()]

    def get_config_resources(self, config: dict[str, str]) -> list[ToolResource]:
        return [ToolResource(name=n, version=v) for n, v in config.items()]

    async def apply(self, resource: ToolResource, exec: Executor) -> Result:
        return await self._mgr.install(resource.name, resource.version, exec)

    async def remove(self, resource_id: str, exec: Executor) -> Result:
        return await self._mgr.remove(resource_id, exec)

    def needs_update(self, host: ToolResource, config: ToolResource) -> bool:
        return host.version != config.version

    def resource_id(self, resource: ToolResource) -> str:
        return resource.name
